New operators

icecream17 opened the pull request:
### Description of the change

Part of #22100 
Also waits

This pr changes the code to use new operators
Currently copying from pr above, progress = src/config.js

### Performance

Code is generally simpler (but haven't tested)

### Applicable issues

#22099

### Blocked until

??= is supported (ES2020), see #21623

### Draft progress

Next file to check: electron-shims.js

icecream17 committed:
Merge pull request #2 from atom/master

Update #1

icecream17 committed:
(Use ?.)

icecream17 committed:
(Use ?. ??)

String.prototype.match returns an array.
Here the regex is "-" ("beta" | "nightly") <number><one or more><end of string>

So if it captures something, the capture is guaranteed to be truthy ("beta" | "nightly")
(Capture 1 = matchArrayResult?.[1])
Therefore it's safe to change || to ?? since there's no difference between the operators

icecream17 committed:
Use ?? ?. (changes != to !== once)

Line 67 is technically different
Line 584 is only different if String.prototype.match is overridden. The original match function always either returns an array or null
Line 885 is also technically different (example: null)
Line 1005 is also technically different (example: {getPath: false})
and also Line 1314
and also Line 1309, but only with a Proxy (example: new Proxy({}, {call(){return 2}}) --> error vs undefined), which is so obscure that I changed it anyways.
Line 1498 is different though because there's an implicit Boolean conversion because of the if statement
paths && paths.length > 0 ---> Boolean(paths) && paths.length > 0
paths?.length > 0 ---> paths.length == null ? false : paths.length > 0

Theoretically, if paths was falsy, but paths.length was set to 1, it would be different. No good reason to do that for any falsy value


this.project is either truthy or null, so some changes based on that were made
Note to self: Comment out `if (this.project)` on line 453

icecream17 committed:
== changed to ===

babelVersionDirectory only changes once (to a truthy value)

icecream17 committed:
change == to ===

The above 
if (!(color instanceof Color)) {
   color = Color.parse(color)
}
means that (typeof color === Color | null)

icecream17 committed:
Use ??= ??, change one != to !==

this.rootNode = null | truthy
node.parentNode = (parent of this, https://dom.spec.whatwg.org/#dom-node-parentnode) where typeof parent = object | null (https://dom.spec.whatwg.org/#concept-tree-parent)

icecream17 committed:
Use ?. ?? ??=

Line 671: Note to self: this.scopedSettingsStore.getAll can be falsy? Hmm... https://github.com/atom/scoped-property-store/blob/master/src/scoped-property-store.coffee#L78
Line 1075 - 1100: these if statements are horrible. They basically say (if not property or property.property something). If property is falsy, it doesn't have property.property (assumption), so it's equivalent.
Line 1283: Note to self: It's impossible for (value == null && value === false)
Line 1600: Again, Color.parse always returns Color | null

icecream17 committed:
fix linting

Is there a way to skip ci?

icecream17 committed:
fix linting

icecream17 reviewed (commented):
(I was trying to clarify my own commit description - ignore this)
```String.prototype.match``` returns ```null``` or an array.
If there's a match, there's a guaranteed array whose [1]nd item is truthy
If there's not a match, it's ```null```.

So ```||``` is equivalent to ```??``` in this case

aminya reviewed (changes requested):
Thanks. Separating the changes is better.

I still think this PR requires
https://github.com/atom/atom/pull/21623

asturur commented on src/atom-environment.js:
```diff
@@ -1306,12 +1300,8 @@ class AtomEnvironment {
-      this.applicationDelegate.setWindowRepresentedFilename(filename);
-    }
+    // eslint-disable-next-line no-unused-expressions
+    this.applicationDelegate.setWindowRepresentedFilename?.(filename);
```
this is not as safe as before, if `setWindowRepresentedFilename` isn't a function this will rise an error.

aminya replied:
I don't think that is accurate. `?.` can also be used for functions.

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Optional_chaining#optional_chaining_with_function_calls

asturur replied:
Yes but if is existing and not a function will throw an error.
![image](https://user-images.githubusercontent.com/1194048/113413518-2c84cc00-93bb-11eb-9a2a-005e347dac83.png)
I was just pointing out that this is not a syntax only change but a behavioural change.

icecream17 committed:
Revert ?. back to typeof check

asturur caught that this wasn't the same. I guess I misunderstood the ?. operator...

asturur commented on src/color.js:
```diff
@@ -110,7 +110,7 @@ module.exports = class Color {
     }
 
-    if (color == null) {
+    if (color === null) {
```
this is also a behavioural change. The previous `==` was catching both null and undefined.
This will not catch an undefined, moving forward with the code and trying to access properties of an undefined.

icecream17 replied:
color will always be either null or an instance of Color

asturur replied:
to me seems that in this case color is an argument of the function, so can be whatever people pass down. Calling the method with no argument before would return false, now will throw an error. Right?

As long as this change is acknowledged and willingful, that is fine.

UziTech replied:
Seems like throwing an error would be the correct thing to do anyway if `color.isEqual` is called without an argument.

icecream17 replied:
Lines 109 to 111 say
```javascript
    if (!(color instanceof Color)) {
      color = Color.parse(color);
    }
```
So actually, `color` has to be `instanceof Color | ReturnType<Color.parse>`
And `Color.parse` never returns `undefined`, therefore `color` can never be `undefined` either

asturur replied:
oh you are right i totally missed that.

A1T0M3CK commented:
https://www.google.com/url?q=https%3A%2F%2Fgithub.com%2Fatom%2Fatom%2Fpull%2F22108%23discussion_r606693998&sa=D&sntz=1&usg=AFQjCNEcZM_YMxj_GurX17josbZCHCLLFQ

On 4/4/21, steven nguyen ***@***.***> wrote:
> @icecream17 commented on this pull request.
>
>
>
>> @@ -110,7 +110,7 @@ module.exports = class Color {
>        color = Color.parse(color);
>      }
>
> -    if (color == null) {
> +    if (color === null) {
>
> Lines 109 to 111 say
> ```javascript
>     if (!(color instanceof Color)) {
>       color = Color.parse(color);
>     }
> ```
> So actually, `color` has to be `instanceof Color | ReturnType<Color.parse>`
> And `Color.parse` never returns `undefined`, therefore `color` can never be
> `undefined` either
>
> --
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly or view it on GitHub:
> https://github.com/atom/atom/pull/22108#discussion_r606693998

A1T0M3CK commented:
View striker=random [color]


On 4/4/21, TorBroser Os ***@***.***> wrote:
> https://www.google.com/url?q=https%3A%2F%2Fgithub.com%2Fatom%2Fatom%2Fpull%2F22108%23discussion_r606693998&sa=D&sntz=1&usg=AFQjCNEcZM_YMxj_GurX17josbZCHCLLFQ
>
> On 4/4/21, steven nguyen ***@***.***> wrote:
>> @icecream17 commented on this pull request.
>>
>>
>>
>>> @@ -110,7 +110,7 @@ module.exports = class Color {
>>        color = Color.parse(color);
>>      }
>>
>> -    if (color == null) {
>> +    if (color === null) {
>>
>> Lines 109 to 111 say
>> ```javascript
>>     if (!(color instanceof Color)) {
>>       color = Color.parse(color);
>>     }
>> ```
>> So actually, `color` has to be `instanceof Color |
>> ReturnType<Color.parse>`
>> And `Color.parse` never returns `undefined`, therefore `color` can never
>> be
>> `undefined` either
>>
>> --
>> You are receiving this because you are subscribed to this thread.
>> Reply to this email directly or view it on GitHub:
>> https://github.com/atom/atom/pull/22108#discussion_r606693998
>

icecream17 committed:
Use ?.

icecream17 committed:
fix linting and use ?. ??

icecream17 committed:
fix linting

icecream17 committed:
Use ?.

icecream17 committed:
Use ??=

icecream17 committed:
Use ?.

icecream17 committed:
Use ?.

It might be possible to replace some of the || with ?? but I can't tell where the code leads

icecream17 commented:
The lint errors with no-unused-expressions and nullish coalescing are fixed in babel 7

icecream17 committed:
Use ??=, remove lint comment

loadSettingsJSON is defined at src/main-process/atom-window.js#L81.\nThe getter always returns a JSON stringified object, so the change is valid

icecream17 committed:
typo

icecream17 committed:
Merge pull request #6 from atom/master

Update from atom/master

icecream17 commented:
I've decided to make a pr with only `??` since that seems to be already supported
